executor/sort.go

```diff
@@ -412,3 +399,210 @@ func (e *TopNExec) Next(goCtx goctx.Context) (Row, error) {
 	e.Idx++
 	return row, nil
 }
+
+// topNChunkHeap implements heap.Interface.
+type topNChunkHeap struct {
+	*TopNExec
+}
+
+// Less implement heap.Interface, but since we mantains a max heap,
+// this function returns true if row i is greater than row j.
+func (h *topNChunkHeap) Less(i, j int) bool {
+	if h.keyChunks != nil {
+		return h.keyChunksGreater(i, j)
+	}
+	return h.keyColumnsGreater(i, j)
+}
+
+func (h *topNChunkHeap) keyChunksGreater(i, j int) bool {
+	keyRowI := h.keyChunks.GetRow(h.rowPtrs[i])
+	keyRowJ := h.keyChunks.GetRow(h.rowPtrs[j])
+	return h.greaterRow(keyRowI, keyRowJ)
+}
+
+func (h *topNChunkHeap) keyColumnsGreater(i, j int) bool {
+	rowI := h.rowChunks.GetRow(h.rowPtrs[i])
+	rowJ := h.rowChunks.GetRow(h.rowPtrs[j])
+	return h.greaterRow(rowI, rowJ)
+}
+
+func (h *topNChunkHeap) greaterRow(rowI, rowJ chunk.Row) bool {
+	for i, colIdx := range h.keyColumns {
+		cmpFunc := h.keyCmpFuncs[i]
+		cmp := cmpFunc(rowI, colIdx, rowJ, colIdx)
+		if h.ByItems[i].Desc {
+			cmp = -cmp
+		}
+		if cmp > 0 {
+			return true
+		} else if cmp < 0 {
+			return false
+		}
+	}
+	return false
+}
+
+func (h *topNChunkHeap) Len() int {
+	return len(h.rowPtrs)
+}
+
+func (h *topNChunkHeap) Push(x interface{}) {
+	// Should never be called.
+}
+
+func (h *topNChunkHeap) Pop() interface{} {
+	h.rowPtrs = h.rowPtrs[:len(h.rowPtrs)-1]
+	// We don't need the popped value, return nil to avoid memory allocation.
+	return nil
+}
+
+func (h *topNChunkHeap) Swap(i, j int) {
+	h.rowPtrs[i], h.rowPtrs[j] = h.rowPtrs[j], h.rowPtrs[i]
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *TopNExec) NextChunk(chk *chunk.Chunk) error {
+	chk.Reset()
+	if !e.fetched {
+		e.totalLimit = int(e.limit.Offset + e.limit.Count)
+		e.Idx = int(e.limit.Offset)
+		err := e.loadChunksUntilTotalLimit()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = e.executeTopN()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.fetched = true
+	}
+	if e.Idx >= len(e.rowPtrs) {
+		return nil
+	}
+	maxChkSize := e.ctx.GetSessionVars().MaxChunkSize
+	for chk.NumRows() < maxChkSize && e.Idx < len(e.rowPtrs) {
+		row := e.rowChunks.GetRow(e.rowPtrs[e.Idx])
+		chk.AppendRow(0, row)
+		e.Idx++
+	}
+	return nil
+}
+
+func (e *TopNExec) loadChunksUntilTotalLimit() error {
+	e.chkHeap = &topNChunkHeap{e}
+	e.rowChunks = chunk.NewList(e.schema.GetTypes(), e.ctx.GetSessionVars().MaxChunkSize)
+	for e.rowChunks.Len() < e.totalLimit {
+		srcChk := e.children[0].newChunk()
+		err := e.children[0].NextChunk(srcChk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if srcChk.NumRows() == 0 {
+			break
+		}
+		e.rowChunks.Add(srcChk)
+	}
+	e.initPointers()
+	e.initCompareFuncs()
+	allColumnExpr := e.buildKeyColumns()
+	if !allColumnExpr {
+		e.buildKeyExprsAndTypes()
+		err := e.buildKeyChunks()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+const topNCompactionfactor = 4
```

Comment:
factor -> Factor

Reply:
Can we make it configurable ?

Reply:
I don't see any reason to modify this value because it doesn't matter that much.
We can always add it later when we found the reason to modify it.